lib/db/schemaupdater.go

```diff
@@ -629,5 +669,304 @@ func (db *schemaUpdater) updateSchemaTo12(_ int) error {
 		return err
 	}
 
+	if err := db.rewriteGlobals(t); err != nil {
+		return err
+	}
+
 	return t.Commit()
 }
+
+func (db *schemaUpdater) rewriteGlobals(t readWriteTransaction) error {
+	it, err := t.NewPrefixIterator([]byte{KeyTypeGlobal})
+	if err != nil {
+		return err
+	}
+	for it.Next() {
+		var vl VersionListDeprecated
+		if err := vl.Unmarshal(it.Value()); err != nil {
+			// If we crashed during an earlier migration, some version
+			// lists might already be in the new format: Skip those.
+			var nvl VersionList
+			if nerr := nvl.Unmarshal(it.Value()); nerr == nil {
+				continue
+			}
+			return err
+		}
+		if len(vl.Versions) == 0 {
+			if err := t.Delete(it.Key()); err != nil {
+				return err
+			}
+		}
+
+		var newVl VersionList
+		newPos := -1
+		oldPos := 0
+		var lastVersion protocol.Vector
+		for _, fv := range vl.Versions {
+			if fv.Invalid {
+				break
+			}
+			oldPos++
+			if lastVersion.Equal(fv.Version) {
+				newVl.RawVersions[newPos].Devices = append(newVl.RawVersions[newPos].Devices, fv.Device)
+				continue
+			}
+			newVl.RawVersions = append(newVl.RawVersions, newFileVersion(fv.Device, fv.Version, false, fv.Deleted))
+			lastVersion = fv.Version
+			newPos++
```

Comment:
`newPos = len(newVl.RawVersions)` before the append avoids the odd initialization to -1 and the risk of somehow getting out of sync in the ++